### PR TITLE
Use tags to support custom resource IDs for VPC, Subnet, and SecurityGroup resources

### PR DESCRIPTION
**Pull request by RobertLucian**
## Motivation

To support the use-case where a given user needs to deploy an IaC infra that makes references to external resources. Those external resources are given a unique resource ID when creating the resource, so attempting to recreate the exact same resources on LocalStack is not possible.

The alternative this PR is proposing is to use a custom tag `_custom_id_` to indicate what the resource ID of a resource ought to be. i.e.:

```bash
awslocal ec2 create-subnet --vpc-id $vpc_id --cidr-block 172.31.112.0/20 --tag-specifications "ResourceType=subnet,Tags=[{Key=_custom_id_,Value=123456789}]"
```

Where `1234556789` is the custom ID that's used for the given subnet.

## Changes

Allows the use of the of `_custom_id_` tag for VPCs, Subnets and SecurityGroups resources to override the UID that's generated for each resource with the custom one associated to `_custom_id_` key.

## TODO

What's left to do:

- [x] Fix custom ID not showing up on the created resource.
- [x] Add tests for when the custom ID tag is attached.

### CODE REVIEW

**Comment by RobertLucian** [comment]
I have read the CLA Document and I hereby sign the CLA.

**Comment by whummer** [comment]
recheck

**Review by dominikschubert — COMMENTED** [review]
General structure LGTM, thanks for jumping on this.

I'll leave it to @viren-nadkarni to sign off on the moto patches :+1: 

Might be good to add a comment to the patches with a short explanation of why the patch is needed, in this case basically explaining the need to assign a specific predefined ID

File: tests/aws/services/ec2/test_ec2.py
Comment (RobertLucian) [reply tests/aws/services/ec2/test_ec2.py:418]: Gotcha. I don't see the autofill on the `aws_client.ec2` resource though. Then again, shouldn't the `aws_client` param have a type hint in order for the autofill to work?

File: localstack-core/localstack/services/ec2/exceptions.py
Comment (RobertLucian) [reply localstack-core/localstack/services/ec2/exceptions.py:67]: Good catch.

File: tests/aws/services/ec2/test_ec2.py
Comment (RobertLucian) [reply tests/aws/services/ec2/test_ec2.py:422]: I like the first option more. I'll go with that option then. Thanks!

File: tests/aws/services/ec2/test_ec2.py
Comment (viren-nadkarni) [inline tests/aws/services/ec2/test_ec2.py:57]: For added cleanup resilience, you can

```suggestion
        try:
            aws_client.ec2.delete_vpc(VpcId=vpc_id)
        except Exception:
            pass
```

File: tests/aws/services/ec2/test_ec2.py
Comment (viren-nadkarni) [inline tests/aws/services/ec2/test_ec2.py:48]: Let's follow PEP8 and avoid camel case for argument names

```suggestion
        cidr_block: str,
        tag_specifications: list[dict],
```

**Review by viren-nadkarni — CHANGES_REQUESTED** [review]

File: tests/aws/services/ec2/test_ec2.py
Comment (RobertLucian) [reply tests/aws/services/ec2/test_ec2.py:57]: Cool!

File: tests/aws/services/ec2/test_ec2.py
Comment (RobertLucian) [reply tests/aws/services/ec2/test_ec2.py:48]: Done.

File: localstack-core/localstack/services/ec2/patches.py
Comment (RobertLucian) [reply localstack-core/localstack/services/ec2/patches.py:6]: Good observation. Moved it to the constants module.

File: localstack-core/localstack/services/ec2/patches.py
Comment (RobertLucian) [reply localstack-core/localstack/services/ec2/patches.py:99]: Agreed. There's no reason why we cannot do this before calling Moto.

File: localstack-core/localstack/services/ec2/patches.py
Comment (viren-nadkarni) [reply localstack-core/localstack/services/ec2/patches.py]: I meant redefining this constant in `localstack.services.ec2.constants` module.

But I suppose this approach of a common constant is fine too. This is becoming a consistent pattern for using deterministic resource IDs.

**Review by viren-nadkarni — APPROVED** [review]
Looks good :100: 

Can be merged after CI is :green_circle:

**Comment by RobertLucian** [comment]
@viren-nadkarni is this a fluke? https://app.circleci.com/pipelines/github/localstack/localstack/27473/workflows/eb7d2a38-fac9-4de9-83d4-f34ae41c9f66/jobs/236458

**Comment by bentsku** [comment]
The CircleCI is a fluke, I believe we'll need to skip that flaky test again. However the 3 new tests are failing here: https://github.com/localstack/localstack/runs/29058138936

**Comment by RobertLucian** [comment]
@bentsku @viren-nadkarni I don't get it. The tests are passing on my laptop, consistently. This should be idempotent. What's being missed on this?

**Comment by bentsku** [comment]
Looks like the child class does not call `super().__init__()`, so the patches are not applied

**Pull request merged**
